src/core/server/saved_objects/migrationsv2/index.ts

```diff
@@ -0,0 +1,240 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/* eslint-disable no-console */
+
+import { ElasticsearchClient } from 'src/core/server/elasticsearch';
+import { gt } from 'semver';
+import chalk from 'chalk';
+import * as Actions from './actions';
+
+export type ControlState =
+  | 'INIT' // Start
+  | 'DONE' // Migration complete
+  | 'FATAL' // An error occurred, migration didn't succeed
+  | 'INIT_NEW_INDICES' // Blank ES cluster, create new kibana indices
+  | 'SET_SOURCE_WRITE_BLOCK' // Set a write block on the source index to prevent any further writes
+  | 'CLONE_SOURCE'; // Create the target index by cloning the source index
+
+type ActionResponse =
+  | Actions.CloneIndexResponse
+  | Actions.FetchAliasesResponse
+  | Actions.SetIndexWriteBlockResponse;
+// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
+export type State = {
+  controlState: ControlState;
```

Comment:
probably `State` should a union of all possible states? for example, `target` doesn't make sense on `INIT` step

Reply:
Yeah that's a good idea, especially when we start adding more stat it will make it clearer what state belongs to which controlState